librina/wrap/c/src/librina-c.cc

```diff
@@ -0,0 +1,468 @@
+}
+
+int flow_dealloc(int fd)
+{
+        DeallocateFlowResponseEvent *        resp = NULL;
+	unsigned int                         seq_num = 0;
+	IPCEvent *                           event = NULL;
+        unsigned int                         port_id = 0;
+        struct flow *                        mein_flow;
+        map <int, struct flow *>::iterator   it;
+
+        it = flows.find(fd);
+        if (it == flows.end())
+                return -1;
+
+        flows.erase(fd);
+
+        mein_flow = it->second;
+        port_id = mein_flow->port_id;
+        delete(mein_flow);
+
+        try {
+                seq_num = ipcManager->requestFlowDeallocation(port_id);
+        } catch (IPCException e) {
+                return -1;
+        }
+
+        event = ipcEventProducer->eventWait();
+        while (event == NULL ||
+               event->eventType != DEALLOCATE_FLOW_RESPONSE_EVENT ||
+               event->sequenceNumber != seq_num) {
+                event = ipcEventProducer->eventWait();
+        }
+
+	resp = dynamic_cast<DeallocateFlowResponseEvent*>(event);
+
+        try {
+                ipcManager->flowDeallocationResult(port_id, resp->result == 0);
+        } catch (IPCException e) {
+                return -1;
+        }
+
+	return resp->result;
+}
+
+int flow_cntl(int fd, int oflags)
+{
+        unsigned int                       port_id = 0;
+        struct flow *                      mein_flow;
+        map <int, struct flow *>::iterator it;
+
+        it = flows.find(fd);
+        if (it == flows.end())
+                return -1;
+
+        mein_flow = it->second;
+        port_id = mein_flow->port_id;
+
+        if (oflags & 00004000)
+                return ipcManager->setFlowOptsBlocking(port_id, false);
+        else
+                return ipcManager->setFlowOptsBlocking(port_id, true);
+}
+
+ssize_t flow_write(int fd, void * buf, size_t count)
+{
+        struct flow *                      mein_flow;
+        map <int, struct flow *>::iterator it;
+        int                                ret = 0;
+
+        if (buf == NULL) {
+                return -1;
+        }
+
+        it = flows.find(fd);
+        if (it == flows.end())
+                return -1;
+
+        mein_flow = it->second;
+
+        try {
+                ret = ipcManager->writeSDU(mein_flow->port_id, buf, count);
+        } catch (IPCException e) {
+                return -1;
+        }
+
+        return ret;
+}
+
+ssize_t flow_read(int fd, void * buf, size_t count)
+
+{
+        struct flow *                      mein_flow;
+        map <int, struct flow *>::iterator it;
+        int                                ret = 0;
+
+        if (buf == NULL) {
+                return -1;
+        }
+
+        it = flows.find(fd);
+        if (it == flows.end())
+                return -1;
+
+        mein_flow = it->second;
+
+        try {
+                ret = ipcManager->readSDU(mein_flow->port_id, buf, count);
+        } catch (IPCException e) {
```

Comment:
This is a bit dangerous. While wrapping C++ code for C it is _very_ advisable to have a catch all clause (```catch(...){}```). An uncatched exception not deriving from ```IPCException``` (e.g. an STL exception), slipping through the library, will directly force the termination of the C program.

It is also preferable to catch the exception by reference ```catch (IPCException& e){}```.

This also applies to the rest of try-catch clauses